subsys/mgmt/fmfu/src/mgmt_fmfu.c

```diff
@@ -0,0 +1,276 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <zephyr.h>
+#include <logging/log.h>
+#include <mgmt/mgmt.h>
+#include <nrf_fmfu.h>
+#include <mgmt/mgmt_fmfu.h>
+#include "cborattr/cborattr.h"
+#include <stats/stats.h>
+
+LOG_MODULE_REGISTER(mgmt_fmfu, CONFIG_MGMT_FMFU_LOG_LEVEL);
+
+/* Define full modem update mgmt spesific SMP return codes */
+#define MGMT_ERR_EMODEM_INVALID_COMMAND 200
+#define MGMT_ERR_EMODEM_FAULT           201
+
+/* Define full modem mgmt commands */
+#define FMFU_MGMT_ID_GET_HASH 0
+#define FMFU_MGMT_ID_UPLOAD 1
+
+/* Define full modem IPC failures*/
+#define NRF_FMFU_RET_COMMAND_FAILED -3
+#define NRF_FMFU_RET_COMMAND_FAULT -4
+
+#define FIRMWARE_SHA_LENGTH 32
+
+/* Struct used for parsing upload requests. */
+struct firmware_packet {
+	uint8_t data[SMP_PACKET_MTU];
+	uint32_t target_address;
+	uint32_t data_len;
+};
+
+static int unpack(struct mgmt_ctxt *ctxt, struct firmware_packet *packet,
+		  bool *whole_file_received)
+{
+	static uint32_t file_length;
+	unsigned long long offset;
```

Comment:
I think I have asked about this earlier, but is there no way we can make these `uint32_t` ?

Reply:
Anyway, state here why they have to be long long so that future me will know :)

Reply:
Added some comment about it. Not sure what to write basically you get compiler warning as the type is not compatible with a struct or uint32_t